site/app/controllers/PollController.php

```diff
@@ -145,22 +145,45 @@ public function addNewPoll() {
                 new RedirectResponse($this->core->buildCourseUrl(['polls']))
             );
         }
+
         $response_count = $_POST["response_count"];
         $responses = [];
         $answers = [];
         $orders = [];
-        for ($i = 0; $i < $response_count; $i++) {
-            if (!isset($_POST["option_id_" . $i]) || !isset($_POST["response_" . $i]) || !isset($_POST["order_" . $i])) {
-                $this->core->addErrorMessage("Error occured in adding poll");
+        $count = 0;
+        for ($i = 0; $count < $response_count; $i++) {
+            if (isset($_POST["option_id_" . $i]) || isset($_POST["response_" . $i]) || isset($_POST["order_" . $i])) {
+                // if at least one of the fields for a response with this index
+                // is set, assert that the rest are also set
+                if (!isset($_POST["option_id_" . $i]) || !isset($_POST["response_" . $i]) || !isset($_POST["order_" . $i])) {
+                    $this->core->addErrorMessage("Error with responses occured in adding poll");
+                    return MultiResponse::RedirectOnlyResponse(
+                        new RedirectResponse($this->core->buildCourseUrl(['polls']))
+                    );
+                }
+                if ($_POST["response_" . $i] === "") {
+                    $this->core->addErrorMessage("Error occured in adding poll: responses must not be left blank");
+                    return MultiResponse::RedirectOnlyResponse(
+                        new RedirectResponse($this->core->buildCourseUrl(['polls']))
+                    );
+                }
+                $responses[$_POST["option_id_" . $i]] = $_POST["response_" . $i];
+                $orders[$_POST["option_id_" . $i]] = $_POST["order_" . $i];
+                if (isset($_POST["is_correct_" . $i]) && $_POST["is_correct_" . $i] == "on") {
+                    $answers[] = $_POST["option_id_" . $i];
+                }
+                $count++;
+            }
+            if ($i === 1000) {
+                // the indices assigned to responses don't have to be consecutive,
+                // so to make sure we don't get into an infinite loop, we stop checking
+                // after an arbitrary large number in case $response_count and the
+                // actual number of responses don't match
+                $this->core->addErrorMessage("Error occured in adding poll with the number of responses provided");
```

Comment:
Why not loop through the $_POST array instead of checking from 0 to 1000? Although I can't imagine why the instructor will need more than 1000 different response types, this solution could break if response_count is greater than 1000. It also seems like it might be difficult to manage in the future. Maybe you can even pass in an array of option_id_i's , response_i, and order_?

Reply:
I tried to change the least amount of code in the way polls currently work to fix the bug in the PR, so I didn't change the way  the responses are stored in the `$_POST` array. If I were to iterate through it to extract the indices that are relevant to the responses, I would probably need to use some regex matching for the key strings. Would that be a cleaner solution?

Reply:
I think that would be a good!

Reply:
The better way to handle this is leverage the fact that PHP can handle array names within forms. For example, doing:

```
<input type="hidden" name="foo[]" value="1" />
<input type="hidden" name="foo[]" value="10" />
```

will give you in PHP:
```
$_POST['foo'] = ['foo' => ['1', '10']];
```

You could then just iterate over the singular array for the non-sequential IDs, and doing regex or have this kind of iterator count check.